addons/knowledge/models/knowledge_article.py

```diff
@@ -0,0 +1,992 @@
+        if parent and not parent.user_can_write:
+            raise AccessError(_('You are not allowed to move this article under this parent.'))
+        before_article = self.browse(before_article_id) if before_article_id else False
+
+        # as base user doesn't have access to members, use sudo to allow access it.
+        article_sudo = self.sudo()
+
+        if before_article:
+            sequence = before_article.sequence
+        else:
+            # get max sequence among articles with the same parent
+            sequence = article_sudo._get_max_sequence_inside_parent(parent_id)
+
+        values = {
+            'parent_id': parent_id,
+            'sequence': sequence
+        }
+        if not parent_id:
+            # If parent_id, the write method will set the internal_permission based on the parent.
+            # If set as root article: if moved to private -> set none; if moved to workspace -> set write
+            values['internal_permission'] = 'none' if private else 'write'
+
+        members_to_remove = self.env['knowledge.article.member']
+        if not parent and private:  # If set private without parent, remove all members except current user.
+            member = article_sudo.article_member_ids.filtered(lambda m: m.partner_id == self.env.user.partner_id)
+            if member:
+                members_to_remove = article_sudo.article_member_ids.filtered(lambda m: m.id != member.id)
+                values.update({
+                    'article_member_ids': [(1, member.id, {
+                        'permission': 'write'
+                    })]
+                })
+            else:
+                members_to_remove = article_sudo.article_member_ids
+                values.update({
+                    'article_member_ids': [(0, 0, {
+                        'partner_id': self.env.user.partner_id.id,
+                        'permission': 'write'
+                    })]
+                })
+
+        article_sudo.write(values)
+        members_to_remove.unlink()
+
+        return True
+
+    def article_create(self, title=False, parent_id=False, private=False):
+        parent = self.browse(parent_id) if parent_id else False
+
+        if parent:
+            if private and parent.category != "private":
+                raise ValidationError(_("Cannot create an article under a non-private parent"))
+            if not parent.user_can_write:
+                raise AccessError(_("Cannot create an article under a parent article you can't write on"))
+            if private and not parent.owner_id == self.env.user:
+                raise AccessError(_("Cannot create an article under a non-owned private article"))
+            private = parent.category == "private"
+
+        values = {
+            'parent_id': parent_id,
+            'sequence': self._get_max_sequence_inside_parent(parent_id)
+        }
+        if not parent:
+            values.update({
```

Comment:
Some this shares common code with the move (at least seems), and strange this is not supported by the create or computed fields directly. Don't like this "private" flag that does magic stuff. Can't it be supported cleanly at model level with inverse on internal permission ?